File: 0545-boundary-of-binary-tree/0545-boundary-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var boundaryOfBinaryTree = function(root) {
    
    let ans = [];
    
    
    let isLeaf = (node) => {
        return node.left == null && node.right == null;
    }
    
    let leftSideView  = (node) => {
        if(node) {
            if(isLeaf(node)) return;
            ans.push(node.val);
            if(node.left) {
                leftSideView(node.left);
            } else if(node.right) {
                leftSideView(node.right);
            }
        }
    }
    let rightSideView  = (node) => {
        if(node) {
            if(isLeaf(node)) return;
            if(node.right) {
                rightSideView(node.right);
            } else if(node.left) {
                rightSideView(node.left);
            }
            ans.push(node.val);
        }
    }
    
    let leafNodes = (node) => {
        if(node) {
            if(isLeaf(node) && node != root) {
                ans.push(node.val);
            }
            leafNodes(node.left);
            leafNodes(node.right);
        }
    }
    
    if(root) {
        ans.push(root.val);
        if(root.left) {
            leftSideView(root.left);
        }
        leafNodes(root);
        if(root.right) {
            rightSideView(root.right);
        }
    }
    
    return ans;
    
};

File: 0545-boundary-of-binary-tree/0545-boundary-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public ArrayList<Integer> list;
    public boolean isLeaf(TreeNode node) {
        return node.left == null && node.right == null;
    }
    public void leftView(TreeNode node) {
        if(node != null) {
            if(isLeaf(node)) {
                return;
            }
            list.add(node.val);
            if(node.left != null) {
                leftView(node.left);
            } else if(node.right != null) {
                leftView(node.right);
            }
        }
    }
    public void rightView(TreeNode node) {
        if(node != null) {
            if(isLeaf(node)) {
                return;
            }
            if(node.right != null) {
                rightView(node.right);
            }
            else if(node.left != null) {
                rightView(node.left);
            }
            list.add(node.val);
        }
    }
    
    public void LeafNodes(TreeNode node, TreeNode root) {
        if(node != null) {
            if(isLeaf(node) && node != root) {
                list.add(node.val);
            }
            LeafNodes(node.left,root);
            LeafNodes(node.right,root);
        }
    }
    
    public List<Integer> boundaryOfBinaryTree(TreeNode root) {
        list = new ArrayList<Integer>();
        if(root != null) {
            list.add(root.val);
            leftView(root.left);
            LeafNodes(root, root);
            rightView(root.right);
        }
        
        return list;
    }
}